compiler+runtime/src/cpp/jank/runtime/obj/delay.cpp

```diff
@@ -0,0 +1,103 @@
+#include <jank/runtime/obj/delay.hpp>
+
+namespace jank::runtime
+{
+  obj::delay::static_object(object_ptr const fn)
+    : fn{ fn }
+  {
+    assert(fn);
+    val = nullptr;
+  }
+
+  native_bool obj::delay::equal(object const &o) const
+  {
+    return &o == &base;
+  }
+
+  native_persistent_string obj::delay::to_string() const
+  {
+    fmt::memory_buffer buff;
+    to_string(buff);
+    return native_persistent_string{ reinterpret_cast<native_persistent_string::size_type>(
+                                       buff.data()),
+                                     static_cast<char>(buff.size()) };
+  }
+
+  object_ptr obj::delay::force(object_ptr const &d)
+  {
+    return visit_object(
+      [&]<typename T>(T const typed_s) -> object_ptr {
```

Comment:
Rather than `<typename T>` here, with a param of type `T`, we need to use a different pattern. Let's make the param type `auto`, and then define this type within the lambda: `using T = typename decltype(typed_d)::value_type;`. Also, the convention is to name the typed object after the original object. The original object is `d`, so the param should be `typed_d`. Finally, your lambda is capturing state with the `[&]`, which results in a closure being generated, not a free function. Since you're not actually capturing anything, we can be explicit about that and get better code generation from it.

The reason why we need to do the `using T =` is that the type is going to be a `native_box<>` of our object type. So your `T` will be a `native_box<obj::delay>`. This is a problem, since that type isn't `behavior::derefable`. The resulting code should look like this:

```cpp
return visit_object(
  [](auto const typed_d) -> object_ptr {
    using T = typename decltype(typed_d)::value_type;

    if constexpr(behavior::derefable<T>)
    {
      return typed_d->deref();
    }
    else
    {
      return d;
    }
  },
  d);
```